app/src/main/java/org/thoughtcrime/securesms/home/HomeViewModel.kt

```diff
@@ -31,6 +31,15 @@ class HomeViewModel @Inject constructor(private val threadDb: ThreadDatabase): V
 
     fun tryUpdateChannel() = listUpdateChannel.trySend(Unit)
 
+    override fun onCleared() {
+        super.onCleared()
+
+        for (job in updateJobs) {
+            job.cancel()
+        }
+        updateJobs.clear()
+    }
```

Comment:
The `updateJobs` keep activity context in a globally referenced memory so they have to be cleaned up when viewModel destroys. There's code smells here in terms of how the `LiveData` is used but that's another story.

Reply:
You are right, I didn't see `executor` comes from a `viewModelScope`, I sort of assumed it was one of those Java `ExecutorService` I saw somewhere else.

It is interesting it shows up in the memory reference graph though so I took a closer look at the implementation of `ContentResolver.observeQuery`, which comes from a third party lib: `app.cash.copper.flow`: The snippet of that piece of code is:

```kotlin
@CheckResult
fun ContentResolver.observeQuery(
  uri: Uri,
  projection: Array<String>? = null,
  selection: String? = null,
  selectionArgs: Array<String>? = null,
  sortOrder: String? = null,
  notifyForDescendants: Boolean = false
): Flow<Query> {
  val query = ContentResolverQuery(this, uri, projection, selection, selectionArgs, sortOrder)
  return flow {
    emit(query)

    val channel = Channel<Unit>(CONFLATED)
    val observer = object : ContentObserver(mainThread) {
      override fun onChange(selfChange: Boolean) {
        channel.offer(Unit)
      }
    }

    registerContentObserver(uri, notifyForDescendants, observer)
    try {
      for (item in channel) {
        emit(query)
      }
    } finally {
      unregisterContentObserver(observer)
    }
  }
}
```

I believe the clean up code is not set up correctly,  as `unregisterContentObserver` is never called when debugging the code. It surprises me that it's not called but if I have to DIY I'd use `callbackFlow` so that I can register a cancel hook to proper clean up the observer.

Reply:
Let me come up with something